### PR TITLE
[Data] Update `ExecutionPlan.execute_to_iterator()` to return `RefBundles` instead of `(Block, BlockMetadata)`

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?

Stacked on: 
- [x] https://github.com/ray-project/ray/pull/46455

Followup to https://github.com/ray-project/ray/pull/46369 and https://github.com/ray-project/ray/pull/46455.
Update `ExecutionPlan.execute_to_iterator()` to return `RefBundles` instead of `(Block, BlockMetadata)`, to unify the logic between `RefBundle`s and `Block`s. Also refactor the `iter_batches()` code path accordingly to handle `RefBundle`s instead of raw `Block` and `BlockMetadata`.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/BUILD [hunk 1]
```diff
@@ -21,7 +21,7 @@ py_library(
 py_test_module_list(
     files = glob(["tests/block_batching/test_*.py"]),
     size = "medium",
-    tags = ["team:ml", "exclusive"],
+    tags = ["team:data", "exclusive"],
```
Comment (scottjlee) [inline python/ray/data/BUILD:24]: found that the tests were not being run under data tests, so updated the tag here.

File: python/ray/data/_internal/plan.py [hunk 2]
```diff
@@ -353,20 +352,25 @@ def schema(
...
+                    if metadata.schema is not None and (
+                        metadata.num_rows is None or metadata.num_rows > 0
+                    ):
+                        schema = metadata.schema
+                        break
         elif self.is_read_only():
             # For consistency with the previous implementation, we fetch the schema if
             # the plan is read-only even if `fetch_if_missing` is False.
-            blocks_with_metadata, _, _ = self.execute_to_iterator()
+            # blocks_with_metadata, _, _ = self.execute_to_iterator()
```
Comment (scottjlee) [inline python/ray/data/_internal/plan.py:366]: @scottjlee to remove before merging

**Review by omatthew98 — APPROVED** [review]
lgtm 🚀

File: python/ray/data/_internal/block_batching/iter_batches.py [hunk 3]
```diff
@@ -268,11 +272,12 @@ def prefetch_batches_locally(
         batch_size is None and len(sliding_window) < num_batches_to_prefetch
     ):
         try:
-            next_block_ref_and_metadata = next(block_ref_iter)
+            next_ref_bundle = next(ref_bundles)
+            for block_ref_and_md in next_ref_bundle.blocks:
+                sliding_window.append(block_ref_and_md)
+                current_window_size += block_ref_and_md[1].num_rows
```
Comment (raulchen) [inline python/ray/data/_internal/block_batching/iter_batches.py:278]:
```suggestion
           sliding_window.extend(next_ref_bundle.blocks):
           current_window_size += next_ref_bundle.num_rows()
```

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/execution/interfaces/ref_bundle.py [hunk 4]
```diff
@@ -134,3 +134,13 @@ def _ref_bundles_iterator_to_block_refs_list(
...
+
+
+def _bundle_to_block_md_iterator(
+    ref_bundles: Iterator[RefBundle],
+) -> Iterator[Tuple[ObjectRef[Block], BlockMetadata]]:
+    """Convert an iterator of RefBundles to an iterator of
+    `(Block object reference, corresponding BlockMetadata)`."""
+    for ref_bundle in ref_bundles:
+        for block_ref, metadata in ref_bundle.blocks:
+            yield block_ref, metadata
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/interfaces/ref_bundle.py:146]: This isn't called anywhere. Let's remove it?

**Review by bveeramani — APPROVED** [review]
Nice

File: python/ray/data/_internal/execution/interfaces/ref_bundle.py (hunk 4 above)
Comment (scottjlee) [reply python/ray/data/_internal/execution/interfaces/ref_bundle.py:146]: thanks, removed.

**Pull request merged**